### ISSUE TITLE
[1.12.0-rc4] Impossible to remove a down Master node.

### ISSUE THREAD

**Issue by diegofernandes** [docker/docker#24600]
**Output of `docker version`:**

```
Client:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 04:05:31 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.0-rc4
 API version:  1.24
 Go version:   go1.6.2
 Git commit:   e4a0dbc
 Built:        Wed Jul 13 04:05:31 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 2
 Running: 1
 Paused: 0
 Stopped: 1
Images: 3
Server Version: 1.12.0-rc4
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 30
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: active
 NodeID: d2k2i4wri7hm36jgdqki577m8
 IsManager: Yes
 Managers: 1
 Nodes: 4
 CACertHash: sha256:7310eb967df3e291b57144726175879e92a9815566071663760c6ec68ecc3d64
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-28-generic
Operating System: Ubuntu 16.04 LTS
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 236.4 MiB
Name: master1
ID: 52L6:KUMM:BFGX:NEA3:Q4BQ:WFHH:YQF2:CCHX:RXM3:QCVB:M2ME:AK3T
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Running on VirtualBox

**Steps to reproduce the issue:**
1. Join a node
2. Promote that node
3. Shutdown the node
4. Try to remove the node `docker node rm`
5. Docker demote the node

**Describe the results you received:**
It`s impossible to remove

```
ubuntu@master1:~$ docker node rm 9cmcpdlmzqjltzv61b4y7ufpn
Error response from daemon: rpc error: code = 9 desc = node 9cmcpdlmzqjltzv61b4y7ufpn role is set to manager. It should be demoted to worker for safe removal
```

And It`s impossible to demote

```
docker node demote 9cmcpdlmzqjltzv61b4y7ufpn
Error response from daemon: rpc error: code = 5 desc = can't find manager in raft memberlist
```

**Describe the results you expected:**
Whould be great a `force`option to do that.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Comment by thaJeztah** [comment]
/cc @abronan I think this is the same issue with just 2 managers

@diegofernandes we're working on documentation for this, but basically in a raft, there should always be an odd number of managers (1,3,5 or 7). Having two managers, with one manager down results in a situation that there can *never* be reached consensus, because >50% of the managers need to "agree"

**Comment by ralphtheninja** [comment]
+1 I ran into this as well when playing around

**Comment by abronan** [comment]
Scaling down from 2 to 1 Manager (using `demote`) only works when the 2 nodes are up. This is specific to Raft: whatever you do you need a majority agreeing. In this case with 2 nodes, you end up stuck requiring the 2 nodes to agree on the removal but because the second node is down, you can't.

If the Manager you want to remove is down on a 2 Managers cluster, you have two solutions:

- Bring the node back up and remove it gracefully.
- Restart the first node using `--force-new-cluster` as there is no majority left to process the **removal**.

If you had 3 Managers running and you shut down one of the Managers, it would let the removal be processed because a majority would agree (2/3).

That's also why we recommend using a minimum of 3 Managers with a recommended number of 5 (and the number should be odd: 3/5/7/9, etc. to cope with network partition scenarios).

Hope this is helpful, but I think this is more of a documentation problem.

**Comment by abronan** [comment]
@diegofernandes Looking into the errors:

```
Error response from daemon: rpc error: code = 5 desc = can't find manager in raft memberlist
```

Seems strange to me. Was this a typo on the node ID?

It should give:

```
Error response from daemon: rpc error: code = 4 desc = context deadline exceeded
```

**Comment by thaJeztah** [comment]
Also see https://github.com/docker/docker/issues/24407, and https://github.com/docker/docker/issues/24109

**Comment by diegofernandes** [comment]
@abronan this node was very strange to me too, I have tried to promote but in the node list It was marked as a worker.

Could be something with the hostname?

**Comment by diegofernandes** [comment]
@abronan I tried to bring back the node but no success. I don`t if I did correct in my env. 
I Just start the machine and wait to the node come back but it didn`t happen, so I try a swarm join and it appears as a new node

**Comment by npcode** [comment]
I have a similar problem. One of my managers is down and I can't remove it.

```
$ docker node ls
ID                           HOSTNAME  MEMBERSHIP  STATUS  AVAILABILITY  MANAGER STATUS
52oklvepdg52anqeh80ittgh2    aws3      Accepted    Ready   Active        Reachable
aqt1wn4aarb34rqyherksb6p7    aws4      Accepted    Ready   Active        Reachable
d05oirzqq9lteefzbsdepjzuu *  aws1      Accepted    Ready   Active        Leader
db1vlkevh5jf27al16pr7ff8w    aws5      Accepted    Ready   Active
dvu6kyngrzblajw9unl1jnnw3    aws2      Accepted    Down    Active        Unreachable
$ docker node rm aws2
Error response from daemon: rpc error: code = 9 desc = node dvu6kyngrzblajw9unl1jnnw3 is a cluster manager and is part of the quorum. It must be demoted to worker before removal
$ docker node demote aws2
Manager aws2 demoted in the swarm.
$ docker node rm aws2
Error response from daemon: rpc error: code = 9 desc = node dvu6kyngrzblajw9unl1jnnw3 is a cluster manager and is part of the quorum. It must be demoted to worker before removal
```

**Comment by abronan** [comment]
This should be fixed by docker/swarmkit#1182 and docker/swarmkit#1184

One is still under review. But we'll confirm and close after vendoring the changes in a new version of swarmkit.

**Comment by abronan** [comment]
Should be fixed as part of the swarmkit vendoring in #24620. Please let us know if you still encounter any issue with the node promotion/demotion process by opening a new issue. Thanks.
